app/upgrades/v12/upgrade_test.go

```diff
@@ -0,0 +1,90 @@
+package v12_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+
+	"github.com/osmosis-labs/osmosis/v11/app/apptesting"
+)
+
+type UpgradeTestSuite struct {
+	apptesting.KeeperTestHelper
+}
+
+func (suite *UpgradeTestSuite) SetupTest() {
+	suite.Setup()
+}
+
+func TestKeeperTestSuite(t *testing.T) {
+	suite.Run(t, new(UpgradeTestSuite))
+}
+
+const dummyUpgradeHeight = 5
+
+// note that this test does not perfectly test state migration, as pre_upgrade function
+// which includes setting up tests by creating new pools would also create twap records
+// automatically with this binary. The goal of this test is to test that the upgrade handler
+// does not panic upon upgrade.
+// Detailed state migration tests are placed within the twap keeper.
```

Comment:
This is an important point. I think this is one of those things that just gets manually tested, because implementing it in say, e2e test suite, wouldn't make sense since it will only ever happen once. Since we have significant state on our testnet, if migration works there I would say we are good to go.

Reply:
Wait I thought e2e test is for helping us with upgrade creation / things across versions

Reply:
So e2e test will test this in the sense that, the state migration will happen and we just assume that if it doesn't panic then the state migration happened correctly. What I meant was, we wouldn't implement some e2e test that checks that the state migration ACTUALLY occurred correctly, but (on testnet say) we should manually inspect prior existing pools to ensure that post migration there are no oddities. I would also feel better seeing testnet, which has many pools and lots of existing state, go through this migration to feel 100%

Reply:
I guess what I am saying is, e2e will test this, but maybe its an item we highlight as needing another look QA wise when testnet gets upgraded!